target/linux/ramips/dts/mt7620a_hiwifi_hc5x61.dtsi

```diff
@@ -69,10 +88,23 @@
 				read-only;
 			};
 
-			bdinfo: partition@fe0000 {
+			partition@fe0000 {
+				compatible = "nvmem-cells";
 				label = "bdinfo";
 				reg = <0xfe0000 0x10000>;
 				read-only;
+
+				nvmem-layout {
+					compatible = "fixed-layout";
+					#address-cells = <1>;
+					#size-cells = <1>;
+
+					macaddr_bdinfo_18a: macaddr@18a {
+						compatible = "mac-base";
+						reg = <0x18a 0x11>;
```

Comment:
@rmilecki Hi! I found the commit https://github.com/openwrt/openwrt/commit/5356462ce54734afd32fce83ad118c58cfeb2a55 ("Revert "kernel: switch back 5.15 to fw_devlink=permissive"") broke the `mac-base` NVMEM cell support on ath79 and ramips target. Both wireless and ethernet devices can't be probed. I can confirm that reverting it will make things work again.
<details>
<summary>failed bootlog</summary>

```
[    0.000000] Linux version 5.15.127 (db@Aspire-V-Nitro) (mipsel-openwrt-linux-musl-gcc (OpenWrt GCC 12.3.0 r23755+9-baed68bac8) 12.3.0, GNU ld (GNU Binutils) 2.40.0) #0 Wed Aug 23 23:54:15 2023
[    0.000000] Board has DDR2
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7620A ver:2 eco:6
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019650 (MIPS 24KEc)
[    0.000000] MIPS: machine is HiWiFi HC5861
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32480
[    0.000000] Kernel command line: console=ttyS0,115200 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Writing ErrCtl register=00004000
[    0.000000] Readback ErrCtl register=00004000
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 120988K/131072K available (5529K kernel code, 598K rwdata, 1200K rodata, 1176K init, 215K bss, 10084K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] CPU Clock: 580MHz
[    0.000000] clocksource: systick: mask: 0xffff max_cycles: 0xffff, max_idle_ns: 583261500 ns
[    0.000000] systick: running - mult: 214748, shift: 32
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6590553264 ns
[    0.000002] sched_clock: 32 bits at 290MHz, resolution 3ns, wraps every 7405115902ns
[    0.008033] Calibrating delay loop... 385.84 BogoMIPS (lpj=1929216)
[    0.074281] pid_max: default: 32768 minimum: 301
[    0.080069] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.087483] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.104954] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.114997] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.122087] pinctrl core: initialized pinctrl subsystem
[    0.128877] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.409923] rt2880_gpio 10000600.gpio: registering 24 gpios
[    0.415611] rt2880_gpio 10000600.gpio: registering 24 irq handlers
[    0.422314] rt2880_gpio 10000688.gpio: registering 1 gpios
[    0.427881] rt2880_gpio 10000688.gpio: registering 1 irq handlers
[    0.435760] PCI host bridge to bus 0000:00
[    0.439883] pci_bus 0000:00: root bus resource [mem 0x20000000-0x2fffffff]
[    0.446934] pci_bus 0000:00: root bus resource [io  0x10160000-0x1016ffff]
[    0.453867] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    0.461986] pci 0000:00:00.0: [1814:0801] type 01 class 0x060400
[    0.468048] pci 0000:00:00.0: reg 0x10: [mem 0x00000000-0x7fffffff]
[    0.474425] pci 0000:00:00.0: reg 0x14: [mem 0x00000000-0x0000ffff]
[    0.480862] pci 0000:00:00.0: supports D1
[    0.484877] pci 0000:00:00.0: PME# supported from D0 D1 D3hot
[    0.492752] pci 0000:00:00.0: bridge configuration invalid ([bus 00-00]), reconfiguring
[    0.501109] pci 0000:01:00.0: [14c3:7662] type 00 class 0x028000
[    0.507197] pci 0000:01:00.0: reg 0x10: [mem 0x00000000-0x000fffff 64bit]
[    0.514147] pci 0000:01:00.0: reg 0x30: [mem 0x00000000-0x0000ffff pref]
[    0.521006] pci 0000:01:00.0: PME# supported from D0 D3hot D3cold
[    0.529246] pci_bus 0000:01: busn_res: [bus 01-ff] end is updated to 01
[    0.535942] pci_bus 0000:00: busn_res: [bus 00-ff] end is updated to 01
[    0.542745] pci 0000:00:00.0: BAR 0: no space for [mem size 0x80000000]
[    0.549413] pci 0000:00:00.0: BAR 0: failed to assign [mem size 0x80000000]
[    0.556511] pci 0000:00:00.0: BAR 8: assigned [mem 0x20000000-0x200fffff]
[    0.563406] pci 0000:00:00.0: BAR 9: assigned [mem 0x20100000-0x201fffff pref]
[    0.570765] pci 0000:00:00.0: BAR 1: assigned [mem 0x20200000-0x2020ffff]
[    0.577672] pci 0000:01:00.0: BAR 0: assigned [mem 0x20000000-0x200fffff 64bit]
[    0.585125] pci 0000:01:00.0: BAR 6: assigned [mem 0x20100000-0x2010ffff pref]
[    0.592450] pci 0000:00:00.0: PCI bridge to [bus 01]
[    0.597508] pci 0000:00:00.0:   bridge window [mem 0x20000000-0x200fffff]
[    0.604404] pci 0000:00:00.0:   bridge window [mem 0x20100000-0x201fffff pref]
[    0.612355] clocksource: Switched to clocksource MIPS
[    0.618989] NET: Registered PF_INET protocol family
[    0.624282] IP idents hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.632539] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.641040] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.648917] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.656717] TCP bind hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.663882] TCP: Hash tables configured (established 1024 bind 1024)
[    0.670526] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.677208] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.684760] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.690555] PCI: CLS 0 bytes, default 32
[    0.694838] rt-timer 10000100.timer: maximum frequency is 1220Hz
[    0.706261] workingset: timestamp_bits=14 max_order=15 bucket_order=1
[    0.719289] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.725271] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.738291] gpio-export gpio_export: 2 gpio(s) exported
[    0.744446] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.752073] printk: console [ttyS0] disabled
[    0.756551] 10000c00.uartlite: ttyS0 at MMIO 0x10000c00 (irq = 20, base_baud = 2500000) is a Palmchip BK-3103
[    0.766648] printk: console [ttyS0] enabled
[    0.766648] printk: console [ttyS0] enabled
[    0.775104] printk: bootconsole [early0] disabled
[    0.775104] printk: bootconsole [early0] disabled
[    0.798902] spi spi0.0: force spi mode3
[    0.804285] spi-nor spi0.0: w25q128 (16384 Kbytes)
[    0.809352] 7 fixed-partitions partitions found on MTD device spi0.0
[    0.816280] Creating 7 MTD partitions on "spi0.0":
[    0.821191] 0x000000000000-0x000000030000 : "u-boot"
[    0.829043] 0x000000030000-0x000000040000 : "hw_panic"
[    0.835742] 0x000000040000-0x000000050000 : "factory"
[    0.844245] 0x000000050000-0x000000fc0000 : "firmware"
[    0.851037] 2 uimage-fw partitions found on MTD device firmware
[    0.857187] Creating 2 MTD partitions on "firmware":
[    0.862265] 0x000000000000-0x000000230000 : "kernel"
[    0.870567] 0x000000230000-0x000000f70000 : "rootfs"
[    0.876947] mtd: setting mtd5 (rootfs) as root device
[    0.883395] 1 squashfs-split partitions found on MTD device rootfs
[    0.889758] 0x000000600000-0x000000f70000 : "rootfs_data"
[    0.896711] 0x000000fc0000-0x000000fe0000 : "oem"
[    0.904843] 0x000000fe0000-0x000000ff0000 : "bdinfo"
[    0.911647] 0x000000ff0000-0x000001000000 : "backup"
[    0.928953] rt2880_wdt 10000120.watchdog: Initialized
[    0.935903] NET: Registered PF_INET6 protocol family
[    0.967120] Segment Routing with IPv6
[    0.970997] In-situ OAM (IOAM) with IPv6
[    0.975259] NET: Registered PF_PACKET protocol family
[    0.980511] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
[    0.993783] 8021q: 802.1Q VLAN Support v1.8
[    1.012461] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    1.026114] Freeing unused kernel image (initmem) memory: 1176K
[    1.032168] This architecture does not have kernel memory protection.
[    1.038783] Run /sbin/init as init process
[    1.826976] init: Console is alive
[    1.831176] init: - watchdog -
[    3.520009] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    3.606447] usbcore: registered new interface driver usbfs
[    3.612239] usbcore: registered new interface driver hub
[    3.617857] usbcore: registered new device driver usb
[    3.634207] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    3.643190] fsl-ehci: Freescale EHCI Host controller driver
[    3.651950] ehci-platform: EHCI generic platform driver
[    3.667919] phy phy-usbphy.0: remote usb device wakeup disabled
[    3.674003] phy phy-usbphy.0: UTMI 16bit 30MHz
[    3.678540] ehci-platform 101c0000.ehci: EHCI Host Controller
[    3.684445] ehci-platform 101c0000.ehci: new USB bus registered, assigned bus number 1
[    3.692697] ehci-platform 101c0000.ehci: irq 26, io mem 0x101c0000
[    3.722372] ehci-platform 101c0000.ehci: USB 2.0 started, EHCI 1.00
[    3.730042] hub 1-0:1.0: USB hub found
[    3.735009] hub 1-0:1.0: 1 port detected
[    3.747661] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    3.756016] ohci-platform: OHCI generic platform driver
[    3.761788] ohci-platform 101c1000.ohci: Generic Platform OHCI controller
[    3.768852] ohci-platform 101c1000.ohci: new USB bus registered, assigned bus number 2
[    3.777195] ohci-platform 101c1000.ohci: irq 26, io mem 0x101c1000
[    3.857518] hub 2-0:1.0: USB hub found
[    3.862407] hub 2-0:1.0: 1 port detected
[    3.919317] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    3.937989] init: - preinit -
[    5.335783] random: jshn: uninitialized urandom read (4 bytes read)
[    5.519647] random: jshn: uninitialized urandom read (4 bytes read)
/bin/board_detect: /etc/board.d/02_network: line 208: arithmetic syntax error
[    5.723712] random: jshn: uninitialized urandom read (4 bytes read)
Failed to connect to the switch. Use the "list" command to see which switches are available.
Failed to connect to the switch. Use the "list" command to see which switches are available.
Failed to connect to the switch. Use the "list" command to see which switches are available.
Failed to connect to the switch. Use the "list" command to see which switches are available.
[    6.507790] random: procd: uninitialized urandom read (4 bytes read)
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    8.729696] mount_root: jffs2 not ready yet, using temporary tmpfs overlay
[    8.741968] urandom-seed: Seed file not found (/etc/urandom.seed)
[    8.819236] procd: - early -
[    8.822791] procd: - watchdog -
[    9.626397] procd: - watchdog -
[    9.630361] procd: - ubus -
[    9.760471] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.788399] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.802635] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.820334] procd: - init -
Please press Enter to activate this console.
[   10.921782] random: jshn: uninitialized urandom read (4 bytes read)
[   10.989049] random: ubusd: uninitialized urandom read (4 bytes read)
[   11.013701] random: ubus: uninitialized urandom read (4 bytes read)
[   11.264685] kmodloader: loading kernel modules from /etc/modules.d/*
[   12.057855] Bridge firewalling registered
[   12.073800] Loading modules backported from Linux version v6.1.24-0-g0102425ac76b
[   12.081472] Backport generated by backports.git v5.15.92-1-44-gd6ea70fafd36
[   12.295946] xt_time: kernel timezone is -0000
[   12.833200] PPP generic driver version 2.4.2
[   12.853887] NET: Registered PF_PPPOX protocol family
[   12.926509] kmodloader: done loading kernel modules from /etc/modules.d/*
[   14.393012] urngd: v1.0.2 started.
[   17.758277] random: crng init done
[   17.761783] random: 33 urandom warning(s) missed due to ratelimiting
[   56.833079] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[   56.840173] jffs2_build_filesystem(): unlocking the mtd device...
[   56.840181] done.
[   56.848517] jffs2_build_filesystem(): erasing all blocks after the end marker...
[   93.747129] done.
[   93.756810] jffs2: notice: (1930) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[   93.956785] overlayfs: upper fs does not support tmpfile.
```
</details>

Reply:
@pepe2k I saw you reported a strange nvmem layout behavior on http://lists.openwrt.org/pipermail/openwrt-devel/2023-August/041450.html. Can you try if revert commit https://github.com/openwrt/openwrt/commit/5356462ce54734afd32fce83ad118c58cfeb2a55 can fix it?

Reply:
@DragonBluep 

> Can you try if revert commit [5356462](https://github.com/openwrt/openwrt/commit/5356462ce54734afd32fce83ad118c58cfeb2a55) can fix it?

With 5356462ce54734afd32fce83ad118c58cfeb2a55 reverted, everything works as expected, good catch!

Cheers,
Piotr

Reply:
@rmilecki ^

Reply:
@mpratt14: dropping my `fw_devlink` hack revealed another issue reported here.

Should I bring back my hack? Or should we somehow proceed with your #11302?

Reply:
I'd be happy to rebase my PR (and without the DT commit) and we can try to see if it handles this case as well

Reply:
#11302 is rebased and ready for testing again

note the last commit "strict as default". if you still have issues try dropping that last commit

Reply:
> #11302 is rebased and ready for testing again
> 
> note the last commit "strict as default". if you still have issues try dropping that last commit

System didn't boot.
```
[    0.000000] Board has DDR2
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7620A ver:2 eco:6
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019650 (MIPS 24KEc)
[    0.000000] MIPS: machine is HiWiFi HC5861
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32480
[    0.000000] Kernel command line: console=ttyS0,115200 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Writing ErrCtl register=00004000
[    0.000000] Readback ErrCtl register=00004000
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 120924K/131072K available (5531K kernel code, 604K rwdata, 1200K rodata, 1232K init, 215K bss, 10148K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] CPU Clock: 580MHz
[    0.000000] clocksource: systick: mask: 0xffff max_cycles: 0xffff, max_idle_ns: 583261500 ns
[    0.000000] systick: running - mult: 214748, shift: 32
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6590553264 ns
[    0.000002] sched_clock: 32 bits at 290MHz, resolution 3ns, wraps every 7405115902ns
[    0.007673] Calibrating delay loop... 385.84 BogoMIPS (lpj=1929216)
[    0.073633] pid_max: default: 32768 minimum: 301
[    0.079210] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.086295] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.103424] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.113015] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.119796] pinctrl core: initialized pinctrl subsystem
[    0.126348] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.153139] rt2880_gpio 10000600.gpio: registering 24 gpios
[    0.158636] rt2880_gpio 10000600.gpio: registering 24 irq handlers
[    0.165053] rt2880_gpio 10000688.gpio: registering 1 gpios
[    0.170426] rt2880_gpio 10000688.gpio: registering 1 irq handlers
[    0.178330] clocksource: Switched to clocksource MIPS
[    0.184849] NET: Registered PF_INET protocol family
[    0.189886] IP idents hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.197785] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.205940] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.213439] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.220871] TCP bind hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.227707] TCP: Hash tables configured (established 1024 bind 1024)
[    0.234087] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.240477] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.247701] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.253260] PCI: CLS 0 bytes, default 32
[    0.257252] rt-timer 10000100.timer: maximum frequency is 1220Hz
[    0.268283] workingset: timestamp_bits=14 max_order=15 bucket_order=1
[    0.281199] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.286827] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.299669] gpio-export gpio_export: 2 gpio(s) exported
[    0.305466] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.316844] rt2880_wdt 10000120.watchdog: Initialized
[    0.323538] NET: Registered PF_INET6 protocol family
[    0.356202] Segment Routing with IPv6
[    0.359890] In-situ OAM (IOAM) with IPv6
[    0.363764] NET: Registered PF_PACKET protocol family
[    0.368734] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
[    0.381292] 8021q: 802.1Q VLAN Support v1.8
[    0.389019] Warning: unable to open an initial console.
[    0.396197] /dev/root: Can't open blockdev
[    0.400233] VFS: Cannot open root device "(null)" or unknown-block(0,0): error -6
[    0.407437] Please append a correct "root=" boot option; here are the available partitions:
[    0.415581] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)
[    0.423589] Rebooting in 1 seconds..
```

Reply:
sorry, is that 5.15 or 6.1?

Reply:
I'm using the kernel 5.15.

Reply:
I added a backport patch

Reply:
my PR should be fine now, this time I tested myself first, sorry for the trouble